libsession/src/main/java/org/session/libsession/messaging/sending_receiving/MessageReceiverHandler.kt

```diff
@@ -188,173 +200,293 @@ fun MessageReceiver.handleVisibleMessage(message: VisibleMessage, proto: SignalS
+        disableLocalGroupAndUnsubscribe(groupPublicKey, groupID, userPublicKey)
+    } else {
+        val isCurrentUserAdmin = admins.contains(userPublicKey)
+        storage.updateMembers(groupID, newMembers.map { Address.fromSerialized(it) })
+        if (isCurrentUserAdmin) {
+            MessageSender.generateAndSendNewEncryptionKeyPair(groupPublicKey, newMembers)
+        }
+    }
+    val (contextType, signalType) =
+            if (senderLeft) SignalServiceProtos.GroupContext.Type.QUIT to SignalServiceGroup.Type.QUIT
+            else SignalServiceProtos.GroupContext.Type.UPDATE to SignalServiceGroup.Type.UPDATE
+
+    storage.insertIncomingInfoMessage(context, senderPublicKey, groupID, contextType, signalType, name, members, admins)
+}
+
+private fun MessageReceiver.handleClosedGroupMemberLeft(message: ClosedGroupControlMessage) {
+    val context = MessagingConfiguration.shared.context
+    val storage = MessagingConfiguration.shared.storage
+    val senderPublicKey = message.sender ?: return
+    val userPublicKey = storage.getUserPublicKey()!!
+    if (senderPublicKey == userPublicKey) { return } // Check the user leaving isn't us, will already be handled
+    val kind = message.kind!! as? ClosedGroupControlMessage.Kind.MembersAdded ?: return
+    val groupPublicKey = message.groupPublicKey ?: return
+    val groupID = GroupUtil.doubleEncodeGroupID(groupPublicKey)
+    val group = storage.getGroup(groupID) ?: run {
+        Log.d("Loki", "Ignoring closed group info message for nonexistent group.")
+        return
+    }
+    val name = group.title
+    // Check common group update logic
+    val members = group.members.map { it.serialize() }
+    val admins = group.admins.map { it.toString() }
+    if (!isValidGroupUpdate(group, message.sentTimestamp!!, senderPublicKey)) {
+        return
+    }
+    // If admin leaves the group is disbanded
+    val didAdminLeave = admins.contains(senderPublicKey)
+    val updatedMemberList = members - senderPublicKey
+
+    if (didAdminLeave) {
+        disableLocalGroupAndUnsubscribe(groupPublicKey, groupID, userPublicKey)
+    } else {
+        val isCurrentUserAdmin = admins.contains(userPublicKey)
+        storage.updateMembers(groupID, updatedMemberList.map { Address.fromSerialized(it) })
+        if (isCurrentUserAdmin) {
+            MessageSender.generateAndSendNewEncryptionKeyPair(groupPublicKey, updatedMemberList)
+        }
+    }
+    storage.insertIncomingInfoMessage(context, senderPublicKey, groupID, SignalServiceProtos.GroupContext.Type.QUIT, SignalServiceGroup.Type.QUIT, name, members, admins)
+}
+
+private fun isValidGroupUpdate(group: GroupRecord,
+                               sentTimestamp: Long,
+                               senderPublicKey: String): Boolean  {
+    val oldMembers = group.members.map { it.serialize() }
+    // Check that the message isn't from before the group was created
+    if (group.createdAt > sentTimestamp) {
```

Comment:
So instead of `group.createdAt` you actually want to compare to the sent timestamp of the closed group control message that created the group for you :) . You want all updates since you were added to the group basically, not just the updates since you received the message that added you to the group.